README.md
DOCKER를 이용한 DB, Block Chain, WEB SERVER 구조입니다.
DOCKER-COMPOSSE를 이용하여 각 컨테이너 관리를 하였습니다.

DB 
    mysql 8.0.20

SERVER
    NodeJS + Express 를 이용하여 서버를 구축하였습니다.
    Front-End 구성은 HTML,CSS,Javascript를 이용하였습니다.

Block Chain
    GO-Ethereum을 이용하여 Private NetWork를 구축하였습니다.
        personal.unlockaccount(eth.account[0]) // Password : 1234
        miner.start(2)

Network는 bridge연결로 되어있습니다.
    
docker-compose -f .\total-node.yml up
http://localhost:3000


geth_data/history

eth.accounts
miner.start(2)
miner.stop()
miner.start(2)
exirt
eth.accounts
eth.coinbase
eth.accounts[0]
miner.start(1)
miner.stop()
miner.start(1)
miner.stop()
miner.start(1)
miner.stop()
miner.start(1)
miner.stop()

PPserver/public/pymodule/test.py
for i in range(2, 10):
    for j in range(1, 10):
      print ( str(i) + " x " + str(j) + " = " + str(i*j) )

PPserver/sc.js
export const CONTACT_ADDRESS = '0xec8ad9C48974175612a5eB31427fBcdf2020BBE7'
export const owner = '0x3b3c85f2C1b1001c48CfE7c5E4445B0b9CcF5eEA'

export const ABI = [
	{
		"constant": false,
		"inputs": [],
		"name": "kill",
		"outputs": [],
		"payable": false,
		"stateMutability": "nonpayable",
		"type": "function"
	},
	{
		"constant": false,
		"inputs": [
			{
				"name": "name",
				"type": "string"
			},
			{
				"name": "addr",
				"type": "string"
			},
			{
				"name": "hash",
				"type": "string"
			}
		],
		"name": "addvideo",
		"outputs": [
			{
				"name": "",
				"type": "uint256"
			}
		],
		"payable": false,
		"stateMutability": "nonpayable",
		"type": "function"
	},
	{
		"constant": true,
		"inputs": [],
		"name": "vidnum",
		"outputs": [
			{
				"name": "",
				"type": "uint256"
			}
		],
		"payable": false,
		"stateMutability": "view",
		"type": "function"
	},
	{
		"constant": true,
		"inputs": [],
		"name": "owner",
		"outputs": [
			{
				"name": "",
				"type": "address"
			}
		],
		"payable": false,
		"stateMutability": "view",
		"type": "function"
	},
	{
		"constant": true,
		"inputs": [
			{
				"name": "",
				"type": "uint256"
			}
		],
		"name": "VS",
		"outputs": [
			{
				"name": "addr",
				"type": "address"
			},
			{
				"name": "videohash",
				"type": "string"
			},
			{
				"name": "videoname",
				"type": "string"
			},
			{
				"name": "videoaddr",
				"type": "string"
			},
			{
				"name": "transaction",
				"type": "bool"
			}
		],
		"payable": false,
		"stateMutability": "view",
		"type": "function"
	},
	{
		"inputs": [],
		"payable": false,
		"stateMutability": "nonpayable",
		"type": "constructor"
	}
];

PPserver/public/pages/main.js
let searchBar = document.getElementById('search-bar')

//해당 함수 수정 예정
function search() {
    let str = "";
    let val = searchBar.value;
    // console.log(val)
    let n = 0;

    while (n < val.length) {
        if (val[n] == ' ')
            str += '+'
        else
            str += val[n];

        n = n + 1;
    }
    // console.log(str)
   // window.open("https://www.youtube.com/results?search_query=" + str);

}

function mapsearch() {
  let str = "";
  let val = searchBar.value;
  // console.log(val)
  let n = 0;

  while (n < val.length) {
      if (val[n] == ' ')
          str += '+'
      else
          str += val[n];

      n = n + 1;
  }
  // console.log(str)
 // window.open("https://www.youtube.com/results?search_query=" + str);

}


let mainContent = document.getElementById('main-content');
let boardContent = document.getElementById('board-content');

const contents = Array(4).fill().map(() => ({
  title: '202X년 0X월 XX일 오전 7시 인천 연수구 아암대로 중 추돌사고',
  thumbnail: '1.png',
}))


// 데이터셋을 구현하기 전 js파일 내에 임의로 생성 //
contents[1].title = '2022년 0X월 XX일 오전 9시 안산 상록구 중앙동 길거리 집단폭행'
contents[1].thumbnail = '2.png'

contents[2].title = '2022년 0X월 XX일 오후 4시 경부고속도로 천안IC부근 뺑소니'
contents[2].thumbnail = '3.png'

contents[3].title = '2022년 0X월 XX일 오전 11시 천안 동남구 백석대학교 정문 앞 삼거리 인명사고'
contents[3].thumbnail = '3.png'
///////


contents.forEach(({title, thumbnail}) => {
  mainContent.innerHTML += `
    <div class="grid-item" onclick="location.href='/pages/play.html'">
      <img class="thumbnail" src="../image/${thumbnail}" alt="" />
      <div class="div1" style="">
        <p class="capitalize"><b>${title}</b></p>
      </div>
      <div class="div2"></div>
    </div>
  `;
});

boardContent.innerHTML += `
<div class="board_wrap">
<div class="board_title">
    <strong>리뷰 게시판</strong>
    <p>DE-Mart 에서 경험한 여러분의 후기를 남겨보세요</p>
</div>
<div class="board_list_wrap">
    <div class="board_list">
        <div class="top">
            <div class="num">번호</div>
            <div class="title">제목</div>
            <div class="writer">글쓴이</div>
            <div class="date">작성일</div>
            <div class="count">조회</div>
        </div>
        <div>
            <div class="num">5</div>
            <div class="title"><a href="./pages/view.html">안녕하세요, 4X러 8XXX 차주입니다.</a></div>
            <div class="writer">김XX</div>
            <div class="date">202X.X.XX</div>
            <div class="count">1024</div>
        </div>
        <div>
            <div class="num">4</div>
            <div class="title"><a href="./pages/view.html">오토바이랑 접촉사고났던 보행자입니다.</a></div>
            <div class="writer">황XX</div>
            <div class="date">202X.X.XX</div>
            <div class="count">3572</div>
        </div>
        <div>
            <div class="num">3</div>
            <div class="title"><a href="./pages/view.html">킥보드 운전자랑 부딪혔습니다...</a></div>
            <div class="writer">박XX</div>
            <div class="date">202X.X.XX</div>
            <div class="count">4958</div>
        </div>
        <div>
            <div class="num">2</div>
            <div class="title"><a href="./pages/view.html">XX고등학교 인근에서 학교폭력을 당하던 학생입니다.</a></div>
            <div class="writer">신XX</div>
            <div class="date">202X.X.XX</div>
            <div class="count">7645</div>
        </div>
        <div>
            <div class="num">1</div>
            <div class="title"><a href="./pages/view.html">안녕하세요, 3X바 7XXX 차주입니다.</a></div>
            <div class="writer">홍XX</div>
            <div class="date">202X.X.XX</div>
            <div class="count">5892</div>
        </div>
    </div>
    <div class="board_page">
        <a href="#" class="bt first"></a>
        <a href="#" class="bt prev"></a>
        <a href="#" class="num on">1</a>
        <a href="#" class="num">2</a>
        <a href="#" class="num">3</a>
        <a href="#" class="num">4</a>
        <a href="#" class="num">5</a>
        <a href="#" class="bt next">></a>
        <a href="#" class="bt last">>></a>
    </div>
    <div class="bt_wrap">
        <a href="write.html" class="on">등록</a>
        <!--<a href="#">수정</a>-->
    </div>
</div>
</div>
  `;




PPserver/public/pages/signup.js
// 휴대폰 번호 입력 부분
function changePhone1(){
    const phone1 = document.getElementById("phone1").value // 010
    if(phone1.length === 3){
        document.getElementById("phone2").focus();
    }
}
function changePhone2(){
    const phone2 = document.getElementById("phone2").value // 010
    if(phone2.length === 4){
        document.getElementById("phone3").focus();
    }
}
function changePhone3(){
    const phone3 = document.getElementById("phone3").value // 010
    if(phone3.length === 4){
      document.getElementById("sendMessage").focus();
      document.getElementById("sendMessage").setAttribute("style","background-color:yellow;")
      document.getElementById("sendMessage").disabled = false;
    }
}

// 문자인증+타이머 부분
function initButton(){
  document.getElementById("sendMessage").disabled = true;
  document.getElementById("completion").disabled = true;
  document.getElementById("certificationNumber").innerHTML = "000000";
  document.getElementById("timeLimit").innerHTML = "03:00";
  document.getElementById("sendMessage").setAttribute("style","background-color:none;")
  document.getElementById("completion").setAttribute("style","background-color:none;")
}

let processID = -1;

const getToken = () => {

  // 인증확인 버튼 활성화
  document.getElementById("completion").setAttribute("style","background-color:yellow;")
  document.getElementById("completion").disabled = false;

  if (processID != -1) clearInterval(processID);
  const token = String(Math.floor(Math.random() * 1000000)).padStart(6, "0");
  document.getElementById("certificationNumber").innerText = token;
  let time = 180;
  processID = setInterval(function () {
    if (time < 0 || document.getElementById("sendMessage").disabled) {
      clearInterval(processID);
      initButton();
      return;
    }
    let mm = String(Math.floor(time / 60)).padStart(2, "0");
    let ss = String(time % 60).padStart(2, "0");
    let result = mm + ":" + ss;
    document.getElementById("timeLimit").innerText = result;
    time--;
  }, 50);
};

function checkCompletion(){
  alert("문자 인증이 완료되었습니다.")
  initButton();
  document.getElementById("completion").innerHTML="인증완료"
  document.getElementById("signUpButton").disabled = false;
  document.getElementById("signUpButton").setAttribute("style","background-color:yellow;")
}


// 가입부분 체크

function signUpCheck(){

  let email = document.getElementById("email").value
  let name = document.getElementById("name").value
  let password = document.getElementById("password").value
  let passwordCheck = document.getElementById("passwordCheck").value
  let area = document.getElementById("area").value
  let gender_man = document.getElementById("gender_man").checked
  let gender_woman = document.getElementById("gender_woman").checked
  let check = true;

  // 이메일확인
  if(email.includes('@')){
    let emailId = email.split('@')[0]
    let emailServer = email.split('@')[1]
    if(emailId === "" || emailServer === ""){
      document.getElementById("emailError").innerHTML="이메일이 올바르지 않습니다."
      check = false
    }
    else{
      document.getElementById("emailError").innerHTML=""
    }
  }else{
    document.getElementById("emailError").innerHTML="이메일이 올바르지 않습니다."
    check = false
  }


  // 이름확인
  if(name===""){
    document.getElementById("nameError").innerHTML="이름이 올바르지 않습니다."
    check = false
  }else{
    document.getElementById("nameError").innerHTML=""
  }


  // 비밀번호 확인
  if(password !== passwordCheck){
    document.getElementById("passwordError").innerHTML=""
    document.getElementById("passwordCheckError").innerHTML="비밀번호가 동일하지 않습니다."
    check = false
  }else{
    document.getElementById("passwordError").innerHTML=""
    document.getElementById("passwordCheckError").innerHTML=""
  }

  if(password===""){
    document.getElementById("passwordError").innerHTML="비밀번호를 입력해주세요."
    check = false
  }else{
    //document.getElementById("passwordError").innerHTML=""
  }
  if(passwordCheck===""){
    document.getElementById("passwordCheckError").innerHTML="비밀번호를 다시 입력해주세요."
    check = false
  }else{
    //document.getElementById("passwordCheckError").innerHTML=""
  }


  // 지역선택 확인
  if(area === "지역을 선택하세요."){
    document.getElementById("areaError").innerHTML="지역을 선택해주세요."
    check = false
  }else{
    document.getElementById("areaError").innerHTML=""
  }

  // 성별체크확인
  if(!gender_man && !gender_woman){
    document.getElementById("genderError").innerHTML="성별을 선택해주세요."
    check = false
  }else{
    document.getElementById("genderError").innerHTML=""
  }
  
  if(check){
    document.getElementById("emailError").innerHTML=""
    document.getElementById("nameError").innerHTML=""
    document.getElementById("passwordError").innerHTML=""
    document.getElementById("passwordCheckError").innerHTML=""
    document.getElementById("areaError").innerHTML=""
    document.getElementById("genderError").innerHTML=""
    
    //비동기 처리이벤트
    setTimeout(function() {
      alert("가입이 완료되었습니다.")
      location.href="/";
  },0);
  }
}

PPserver/app.js
const express = require('express');
const Web3 = require('web3'); // web3.js
const app = express()
const multer = require('multer');
const cors = require('cors');
const fs = require('fs');
const path = require('path');
var crypto = require('crypto');
const bodyParser = require('body-parser');
const { resolve } = require('path');
app.use(bodyParser.json())
app.use(bodyParser.urlencoded({ extended: true }));
const mysql = require('mysql2/promise');

var hash = crypto.createHash('md5');
var vihash = "";
const dbConnect = async (num) => {
    try {
        const connection = await mysql.createConnection({
            host: "db",
            user: "root",
            password: "root",
            database: "boardDB",
            charset: "utf8_general_ci"
        });
        console.log("게시판 DB연결 성공");
        var sql = `SELECT hash FROM board WHERE num = ?`
        
        const [rows, fields] = await connection.query(sql,[num]);
        return rows[0].hash;
    } catch (error) {
        console.log(error);
    }
};


const storage = multer.diskStorage({
    destination: function (req, file, cb) {
        cb(null, 'uploads')
    },
    filename: function (req, file, cb) {
        cb(null, file.originalname)
    }
});

const upload = multer({ storage: storage })


/*
    title: '제목테스트',
    state: '장소테스트',
    date: '2022-10-29',
    time: '12:59',
    content: '테스트테스트',
    writer: 'on'
*/
const dbinserttest = async (title,state,datetime,content,url,hash) => {
    try {
        const connection = await mysql.createConnection({
            host: "db",
            user: "root",
            password: "root",
            database: "boardDB",
            charset: "utf8_general_ci"
        });
        var sql =`INSERT INTO Board(title, location ,datetime, contents, url, hash) VALUES (?,?,?,?,?,?);`
        const result123 = await connection.query(sql,[title,state,datetime,content,url,hash]);
        console.log(result123);
    } catch (error) {
        console.log(error);
    }
};
app.post('/uploadtest', upload.single('file'), async (req, res, next) => {
    await filehash(req.file.originalname);
    var t =req.body.title
    var l =req.body.state
    var d =req.body.date + " " + req.body.time
    var con = req.body.content
    var wr = req.body.writer
    var ur = req.file.path
    var 해쉬 = vihash
    dbinserttest(t,l,d,con,ur,해쉬)
    res.json('업로드 완료')
});


app.post('/ver', async (req,res, next) => {
    var vname2 = req.body.vtitle2
    var vn = req.body.vnum2
    var dbhash = await dbConnect(vn); // db에 저장된 해시
    var blockhash = await getcontract(vn) // 블록체인에 저장된 해시 
    console.log(dbhash)
    console.log(blockhash)
    
    if (dbhash === blockhash){
        res.json(vname2 + ', ' + vn + '<div><br />서버에 저장된 정보</div>' + dbhash 
            + '<div><br />블록체인에 저장된 정보</div>' + blockhash + '<div><br />블록체인에 저장된 해시값과 DB에 저장된 해시값이 일치합니다</div><br /> 검증 완료');
    }
    else {
        res.json('일치하지 않습니다.')
    }
})



async function filehash(name) {
    hash = crypto.createHash('md5');
    var fileName = path.join(__dirname + '/uploads', name);
    var input = fs.createReadStream(fileName);
    var h2;
    input.on('readable', function () {
        var data = input.read();
        if (data)
            hash.update(data);
        else {
            var hash2str = hash.digest('hex').toString();
            vihash = hash2str;
            calladdvideo(name, owner, hash2str); // 블록체인등록
        }
    });
}


//172.18.0.3 -> geth컨테이너의 ip주소 컨테이너간 네트워크를 bridge 연결을 해주었기 때문에 확인할 수 있다.


app.use(express.static(__dirname + "/public"));
app.use(cors());

app.get('/', (req, res) => res.sendFile(__dirname + '/public/pages/main.html'));

app.get('/contr', (req, res) => { // test
    dbinsert();
});

/*블록체인과 연동  /contr << getweb3 -> getaccount -> return accounts 계좌불러오기 성공
    컨트랙트와 연동  getweb3 -> getcontract 동기방식
  

*/

const ABI = [
    {
        "constant": false,
        "inputs": [],
        "name": "kill",
        "outputs": [],
        "payable": false,
        "stateMutability": "nonpayable",
        "type": "function"
    },
    {
        "constant": false,
        "inputs": [
            {
                "name": "name",
                "type": "string"
            },
            {
                "name": "addr",
                "type": "string"
            },
            {
                "name": "hash",
                "type": "string"
            }
        ],
        "name": "addvideo",
        "outputs": [
            {
                "name": "",
                "type": "uint256"
            }
        ],
        "payable": false,
        "stateMutability": "nonpayable",
        "type": "function"
    },
    {
        "constant": true,
        "inputs": [],
        "name": "vidnum",
        "outputs": [
            {
                "name": "",
                "type": "uint256"
            }
        ],
        "payable": false,
        "stateMutability": "view",
        "type": "function"
    },
    {
        "constant": true,
        "inputs": [],
        "name": "owner",
        "outputs": [
            {
                "name": "",
                "type": "address"
            }
        ],
        "payable": false,
        "stateMutability": "view",
        "type": "function"
    },
    {
        "constant": true,
        "inputs": [
            {
                "name": "",
                "type": "uint256"
            }
        ],
        "name": "VS",
        "outputs": [
            {
                "name": "addr",
                "type": "address"
            },
            {
                "name": "videohash",
                "type": "string"
            },
            {
                "name": "videoname",
                "type": "string"
            },
            {
                "name": "videoaddr",
                "type": "string"
            },
            {
                "name": "transaction",
                "type": "bool"
            }
        ],
        "payable": false,
        "stateMutability": "view",
        "type": "function"
    },
    {
        "inputs": [],
        "payable": false,
        "stateMutability": "nonpayable",
        "type": "constructor"
    }
];
const contractaddr = '0xec8ad9C48974175612a5eB31427fBcdf2020BBE7'
const owner = '0x3b3c85f2C1b1001c48CfE7c5E4445B0b9CcF5eEA'
const web32 = new Web3(new Web3.providers.HttpProvider('http://172.18.0.4:8545'));
const ethfuntion = new web32.eth.Contract(ABI, contractaddr); //스마트컨트랙트 부르기 ethfuntion.methods.함수명.call()

function getWeb3() {
    const web3 = new Web3(new Web3.providers.HttpProvider('http://172.18.0.4:8545'));
    return web3;
}
async function getAccounts() {   //계정 조회 함수
    try {
        const accounts = await getWeb3().eth.getAccounts();
        console.log(accounts);
        return accounts;
    } catch (e) {
        console.log(e);
        return e;
    }
}

async function getcontract(pram) { // 스마트컨트랙트 내부 조회 함수 pram 은 번호이다= vidnum //getcontract(int) 호출시 
    //addvideo함수작동시 vidnum이라는 변수에 번호를 저장한다  
    try {
        const VS = await ethfuntion.methods.VS(pram).call();
        return VS.videohash;
    } catch (e) {
        console.log(e);
        return e;
    }
}

//(string name, string addr, string hash) 
async function calladdvideo(m1, m2, m3) {   //addvideo 호출 함수  send로 호출해야한다 call과는
    try {
        const ttest = await ethfuntion.methods.addvideo(m1, m2, m3).send({ from: owner, gasPrice: 1e10 });
        console.log(ttest);
        const vidnum = await ethfuntion.methods.vidnum().call();
        console.log("영상 번호는 " + (vidnum - 1));
        hash = crypto.createHash('md5')// hash 객체 초기화
    } catch (e) {
        console.log(e);
        return e;
    }
}




app.listen(3000, () => {
    const dir = "./uploads";
    if (!fs.existsSync(dir)) {
        fs.mkdirSync(dir);
    }
    console.log('포트번호 외부 3000 >>도커 내부3000로 연결');
    console.log('연결후 메인페이지 보여주기');

})
